sound/soc/intel/avs/pcm.c

```diff
@@ -296,7 +296,7 @@ static int avs_dai_hda_be_hw_params(struct snd_pcm_substream *substream,
 	if (data->path)
 		return 0;
 
-	link_stream = substream->runtime->private_data;
+	link_stream = snd_substream_to_hstream(substream);
```

Comment:
... here we assign a 'struct hdac_ext_stream *' pointer.

we shouldn't add more confusion here.

Reply:
I believe this is fine in C. That's how people deal with OOP concepts in the language since its beginnings.

In general, I favor the following approach:
```
struct extension {
    struct base;
    <custom field1>
    ...
    <custom fieldN>
};
```

Candidates in ALSA/ASoC for implementing such approach are many, e.g.: `struct snd_pcm_runtime` and `struct snd_soc_pcm_runtime`.

Also, I have iterated over several options. The problem being addressed here troubles any user of reparenting functionality  when HDAudio LINK stream is involved, including SOF. When fixing this, a developer has to also account for the HDAudio code found in `sound/pci/hda` and, ideally we would like to reuse it without providing any custom implementation.

Reply:
No. I don't want to have to think if I am using the extended version or not. It's should be very very simple and naming conventions do matter.

Reply:
Would adding a second macro - `snd_substream_to_hext_stream` help? That or utilizing `stream_to_hdac_ext_stream` which is already there.

I agree on the naming part, want things clear too. First, we must take care of correctness though. A panic after `dpcm_be_reparent()` is not the right outcome.

Reply:
self-explanatory macros or helpers would be fine. I just want to avoid people having to think on which layer they are using.

Reply:
Introduced additional `snd_substream_to_hext_stream` macro and utilized it each time `hdac_ext_stream` instance is involved. v4 updates skylake-driver too as previous patchset revision found here ignored it.